mempool/cat/pool.go

```diff
@@ -0,0 +1,786 @@
+
+	abci "github.com/tendermint/tendermint/abci/types"
+	"github.com/tendermint/tendermint/config"
+	"github.com/tendermint/tendermint/libs/log"
+	"github.com/tendermint/tendermint/mempool"
+	"github.com/tendermint/tendermint/proxy"
+	"github.com/tendermint/tendermint/types"
+)
+
+// enforce compile-time satisfaction of the Mempool interface
+var _ mempool.Mempool = (*TxPool)(nil)
+
+// Amount of evicted txs to cache. This is low enough so that in patches
+// of high tx throughput, the txpool will be able to quickly recover txs
+// that evicted but not large enough that continual excessive tx load won't
+// take up needless memory.
+const evictedTxCacheSize = 200
+
+var (
+	ErrTxInMempool       = errors.New("tx already exists in mempool")
+	ErrTxAlreadyRejected = errors.New("tx was previously rejected")
+)
+
+// TxPoolOption sets an optional parameter on the TxPool.
+type TxPoolOption func(*TxPool)
+
+// TxPool implemements the Mempool interface and allows the application to
+// set priority values on transactions in the CheckTx response. When selecting
+// transactions to include in a block, higher-priority transactions are chosen
+// first.  When evicting transactions from the mempool for size constraints,
+// lower-priority transactions are evicted first. Transactions themselves are
+// unordered (A map is used). They can be broadcast in an order different from 
+// the order to which transactions are entered. There is no guarantee when CheckTx
+// passes that a transaction has been successfully broadcast to any of its peers.
+//
+// A TTL can be set to remove transactions after a period of time or a number
+// of heights.
+//
+// A cache of rejectedTxs can be set in the mempool config. Transactions that
+// are rejected because of `CheckTx` or other validity checks will be instantly
+// rejected if they are seen again. Committed transactions are also added to
+// this cache. This serves somewhat as replay protection but applications should 
+// implement something more comprehensive
+type TxPool struct {
+	// Immutable fields
+	logger       log.Logger
+	config       *config.MempoolConfig
+	proxyAppConn proxy.AppConnMempool
+	metrics      *mempool.Metrics
+
+	updateMtx            sync.Mutex
+	notifiedTxsAvailable bool
+	txsAvailable         chan struct{} // one value sent per height when mempool is not empty
+	preCheckFn           mempool.PreCheckFunc
+	postCheckFn          mempool.PostCheckFunc
+	height               int64 // the latest height passed to Update
+
+	// Thread-safe cache of rejected transactions for quick look-up
+	rejectedTxCache *LRUTxCache
+	// Thread-safe cache of valid txs that were evicted
+	evictedTxs *EvictedTxCache
+	// Thread-safe list of transactions peers have seen that we have not yet seen
+	seenByPeersSet *SeenTxSet
+
+	// Store of wrapped transactions
+	store *store
+
+	// broadcastCh is an unbuffered channel of new transactions that need to
+	// be broadcasted to peers. Only populated if `broadcast` is enabled
```

Comment:
> Only populated if `broadcast` is enabled

Is `broadcast` a field in the `TxPool`, it does not seem so. Can you please elaborate on it in the comment.

Reply:
`Broadcast` is a field within the `MempoolConfig`. I can elaborate